### ISSUE TITLE
训练管线中 scheduled sampling 可能导致 vision_proj 在 I2T 前被意外更新

### ISSUE THREAD

**Issue by dove667** [jingyaogong/minimind-o#23]
<html><head></head><body><p><span>你好，感谢开源 MiniMind-O。我在使用官方提供的 </span><code><span>llm_768.pth</span></code><span> 和 full 数据集，按照 </span><code><span>trainer/train.sh</span></code><span> 完整复现 Dense 模型时，发现复现权重的文本和语音能力与官方模型接近，但视觉识别与图像分析能力明显弱于官方发布的 </span><code><span>sft_omni_768.pth</span></code><span>。</span></p><p><span>检查训练日志后，确认 T2A、A2A、I2T 以及最终 vision projector 等阶段均正常完成。除长序列 A2A 阶段因显存限制将 </span><code><span>batch_size=32</span></code><span> 改成 </span><code><span>batch_size=16, accumulation_steps=2</span></code><span> 外，其余主要配置均与官方脚本一致，有效 global batch 仍为 128。</span></p><p><span>进一步检查代码和逐阶段权重后，我怀疑当前训练管线可能会在正式 I2T 训练前意外更新或污染 </span><code><span>vision_proj</span></code><span>。</span></p><p><code><span>OmniDataset.apply_scheduled_sampling()</span></code><span> 会将部分 assistant 输入 token 随机替换为整个文本词表中的任意 token：</span></p><pre><code><span>input_ids<span>[8]</span> = torch.where(
    text_mask,
    torch.randint(0, self.text_vocab_size, input_ids<span>[8]</span>.shape),
    input_ids<span>[8]</span>
)</span></code></pre><p><span>当前 tokenizer 词表大小为 6400，而 </span><code><span>&lt;|image_pad|&gt;</span></code><span> 的 token id 为 12。这里虽然保护了原有 image token 不被替换，但没有防止随机生成新的特殊 token。因此，在无图的 T2A/A2A 样本中，普通 assistant token 仍可能被随机替换成 image token。</span></p><p><span>与此同时，无图样本会返回全零的 dummy image：</span></p><pre><code><span>pixel_values = {
    "pixel_values": torch.zeros(1, 3, 256, 256)
}</span></code></pre><p><span>这可能形成如下路径：</span></p><pre><code><span>普通 token 被 scheduled sampling 随机替换为 image token
→ count_vision_proj() 将其识别为图像占位符
→ 对 dummy zero image 计算 SigLIP embedding
→ vision_proj 输出被注入 assistant 序列
→ text/audio loss 反向传播到 vision_proj</span></code></pre><p><span>此外，</span><code><span>MiniMindOmni.forward()</span></code><span> 中还存在：</span></p><pre><code><span>sum(p.sum() for p in self.vision_proj.parameters()) * 0</span></code></pre><p><span>这会让暂未使用的 </span><code><span>vision_proj</span></code><span> 获得非 </span><code><span>None</span></code><span> 的零梯度。在 AdamW 下，即使没有有效视觉监督，也可能继续受到 weight decay。</span></p><p><span>逐阶段检查复现权重时发现，在进入 I2T 之前，</span><code><span>vision_proj</span></code><span> 已经明显偏离标准初始化。例如：

</span></p>
Checkpoint | vision_proj.mlp.0.weight 均值 | vision_proj.mlp.0.bias std |
-- | -- | --
LayerNorm 标准初始化 | 约 1.0 | 0
复现 sft_t2a | 0.6279 | 0.2711
复现 sft_a2a | 0.6253 | 0.2714
复现最终模型 | 0.7291 | 0.2404
官方最终模型 | 0.9393 | 0.0331
<p><span>

其中 A2A projector-only 阶段不会改变 </span><code><span>vision_proj</span></code><span>，因此主要漂移在 T2A full training 阶段就已经出现，而不是由后续 I2T 引起。</span></p><p><span>想请教：</span></p><ol start="1"><li><p><span>官方发布的 </span><code><span>sft_omni_768.pth</span></code><span> 是否严格由当前公开的 </span><code><span>train.sh</span></code><span>、数据和代码版本训练得到？</span></p></li><li><p><span>官方训练时是否在进入 I2T 前重新初始化过 </span><code><span>vision_proj</span></code><span>，或在 T2A/A2A 阶段禁用了 vision encoder / scheduled sampling？</span></p></li><li><p><span>scheduled sampling 是否应该排除 </span><code><span>&lt;|image_pad|&gt;</span></code><span>、</span><code><span>&lt;|audio_pad|&gt;</span></code><span> 等特殊 token？</span></p></li></ol><p><span>
</span></p><p><span>希望作者能帮忙确认以上分析是否正确，如果有误还请指正。谢谢！</span></p></body></html>
